specify/specify7
# Create Custom Django Model Save Function to Allow Overwritting Timestamp Fields

Fixes #1806

Remove the `auto_now_add` attribute for the `timestampCreated` field, and the `auto_now` attribute from the `timestampModified` field.  Replace the functionality of these attributes by overriding the Django model `save()` function with custom logic that allows for these fields to be set in the save function without being overwritten by the current time, but will otherwise use the current time if the field is not specified.  The custom timestamp save function will be used by any model using the timestamp fields, which are most tables in Specify.

The changes being made to the two timestamp fields are detected using the `django-model-utils` library.  This provides a succinct way to detect changes to a fields and provides much more useful Django functionality that could be useful in the future.

This PR implements the timestamp save function as both a separate function set in the `build_models.py` file, but also as an abstract Django model.  In the near future, when the Specify Django models are defined statically, instead of dynamically created from the Specify 6 xml file, this abstract model class will be a better way of overwrite many Django model save functions through inheritance.

### Checklist

      and self-explanatory (or properly documented)
- [ ] UI Testing

### Testing instructions
Upload data in the workbench with `timestampCreated` and `timestampModified` fields.  Then check to make sure those values are indeed the same values that were uploaded, and were not overwritten by the current time.

Make sure to go into preferences -> WorkBench -> No restrictions mode, and set it to true.  That way you will be able to map columns in the workbench to the timestamp fields, otherwise they are hidden, even when checking "Reveal Hidden Form Fields"
